apps/drec-api/src/pods/integrators/integrators.service.ts

```diff
@@ -0,0 +1,194 @@
+    const endDate = DateTime.now()
+      .minus({ minute: 1 })
+      .toUTC()
+      .toFormat('yyyy-MM-dd HH:mm:ss');
+
+    this.logger.debug(`Start date ${startDate} - End date ${endDate}`);
+
+    const devices = await this.deviceService.findMultiple({
+      where: {
+        integrator,
+      },
+    });
+    if (!devices?.length) {
+      return;
+    }
+    const server = this.configService.get<string>('BBOX_SERVER') || '';
+    const username = this.configService.get<string>('BBOX_USERNAME');
+    const password = this.configService.get<string>('BBOX_PASSWORD');
+    const loginForm = new FormData();
+    loginForm.append('username', username);
+    loginForm.append('password', password);
+
+    const authToken = await this.loginBBOX(server, loginForm);
+
+    await Promise.all(
+      devices.map(async (device: DeviceDTO) =>
+        this.storeBBOXenergyReads(
+          server,
+          authToken,
+          device.externalId,
+          startDate,
+          endDate,
+          device.organizationId,
+        ),
+      ),
+    );
+  }
+
+  private loginBBOX(server: string, loginForm: FormData): Promise<string> {
+    return this.httpService
+      .post(`${server}/v1/auth/login`, loginForm, {
+        headers: loginForm.getHeaders(),
+      })
+      .pipe(
+        map(
+          (response) =>
+            response.data.message.login_successful.API_token as string,
+        ),
+        catchError((err) => throwError('Error while logging in to BBOX', err)),
+      )
+      .toPromise();
+  }
+
+  private getBBOXproductReadData(
+    server: string,
+    token: string,
+    productId: string,
+    startDate: string,
+    endDate: string,
+  ): Promise<any> {
+    const requestConfig: AxiosRequestConfig = {
+      headers: { Authorization: `Token token=${token}` },
+      params: {
+        start: startDate,
+        end: endDate,
+        measurement: 'analysis',
+        fields: 'energy_out',
+      },
+    };
+    return this.httpService
+      .get(`${server}/v1/products/${productId}/data`, requestConfig)
+      .pipe(
+        map((resp: any) => {
+          return resp.data.data.energy_out;
+        }),
+        catchError((err) => throwError(err)),
+      )
+      .toPromise();
+  }
+
+  private async storeBBOXenergyReads(
+    server: string,
+    authToken: string,
+    externalId: string,
+    startDate: string,
+    endDate: string,
+    organizationId: number,
+  ): Promise<void> {
+    const energyData = await this.getBBOXproductReadData(
+      server,
+      authToken,
+      externalId,
+      startDate,
+      endDate,
+    );
+
+    if (!energyData?.length) {
+      this.logger.log(
+        `BBOX: No Energy Data found for the interval ${startDate} to ${endDate}`,
+      );
+      return;
+    }
+    const reads: ReadDTO[] = energyData.map((energyValue: string[]) => {
+      const read: ReadDTO = {
+        timestamp: new Date(energyValue[1]),
+        value: +energyValue[0],
```

Comment:
Not a huge fan of casting this way. I'd prefer parseInt, but another question is - are the DTOs/interfaces correct if you need to cast?

Reply:
This data comes from an external API so it needs to be casted.